### DISCUSSION TITLE
Additional SVG Style Elements / Encoded Colors from Array

### DISCUSSION THREAD

**Discussion by sogorman** [chillerlan/php-qrcode#136] [General]
I wanted to check in on the possibility of making a few tweaks to the SVG functionality currently available int the dev-main branch. If we look at `vendor/chillerlan/php-qrcode/tests/samples/vector_sample.svg` as an example we are wanting to create a style where the encoded 1s (dark squares or circles) are either all one hex color or one of two hex colors. 

Lets say a brand has the primary colors of [#463ed7,#4a4699] we would like the encoded data to be a random selection of one of those two colors. Truth be told maybe a random color form an array of N options may be useful to people?

**Comment by codemasher** [comment]
I think the easiest solution would be via the CSS class. The key is the class assignment over here: https://github.com/chillerlan/php-qrcode/blob/2777a0d9125b9f214495ee63e008386942d3f5b0/src/Output/QRMarkupSVG.php#L94-L98
where the line for dark/light could be modified so that it randomly assingns one of a set of classes, like so::

`($M_TYPE & QRMatrix::IS_DARK) === QRMatrix::IS_DARK ? ['dark0', 'dark1'][rand(0, 1)] : 'light'`

Sice this code is still in development I could add a method to prepare the CSS class, so that it won't be necessary to re-implement the whole `paths()` method.

**Reply by codemasher** [reply]
There you go: https://github.com/chillerlan/php-qrcode/commit/851246003c09d834ba045765cb06cae6810354d3
https://github.com/chillerlan/php-qrcode/blob/30693561d7961fc4ac5027bdbe3e665057990f70/src/Output/QRMarkupSVG.php#L113-L119

**Reply by codemasher** [reply]
Wait, this does not (yet) have the desired effect as this css class is for the whole path. You will need to randomly assign the modules to different paths. You see, I'm out of brain juice for today (but at least did a useful commit :).

**Reply by sogorman** [reply]
Thank you for the getCssClass() update. Now able to dictate the CSS classes for the entire path (as you noted).
 
![Screenshot 2022-07-07 160150](https://user-images.githubusercontent.com/40614647/177885438-9b184edb-79e3-41ec-9681-fda11216f0c7.png)

Trying to think of the best way to keep this abstract for use on the modules.

**Reply by codemasher** [reply]
Oh, i thought you wanted to color the data modules randomly (which would be a cool effect btw). Separate colors for the different parts of the matrix, such as the finder and alignment patterns, are already built-in and don't need much changes - [see the GdImage example over here](https://github.com/chillerlan/php-qrcode/blob/19f4fcc0bcea61165e2562ba554f6314a16f7954/examples/image.php) - these work also for SVG. (I know that some things need more detailed documentation - I'll work on that once v5 is ready)

**Reply by sogorman** [reply]
You are correct in looking to stylize the modules randomly between two near colors. just was mentioning that the https://github.com/chillerlan/php-qrcode/commit/851246003c09d834ba045765cb06cae6810354d3 commit and the change to getCssClass().  

Yes the end goal is to have different colored modules to your point probably dictated by the class.

**Reply by codemasher** [reply]
Ah ok, understood. The layer assignment happens over here: https://github.com/chillerlan/php-qrcode/blob/19f4fcc0bcea61165e2562ba554f6314a16f7954/src/Output/QROutputAbstract.php#L163
There you would need to check for the `$M_TYPE === (QRMatrix::M_DATA | QRMatrix::IS_DARK)` and then randomly assign a different layer. The `$M_TYPE_LAYER` is the value of one of the combinations of field values that can be found [over here in `QROutputInterface`](https://github.com/chillerlan/php-qrcode/blob/main/src/Output/QROutputInterface.php#L20-L43). It gets a bit messy from here [in the `paths()` method](https://github.com/chillerlan/php-qrcode/blob/30693561d7961fc4ac5027bdbe3e665057990f70/src/Output/QRMarkupSVG.php#L75-L79) because it requires one (or more) non-existent values, which in turn would need to exist in [the module values array](https://github.com/chillerlan/php-qrcode/blob/main/src/Output/QROutputAbstract.php#L70-L83)... i need to think that through to come up with a clean solution...

**Comment by codemasher** [comment]
Ok, I've pushed [another fix](https://github.com/chillerlan/php-qrcode/commit/1ca995a3bbe576fed7041b2d2836c13bdf1862f9) to avoid errors due to missing module values, which means you can assign `$M_TYPE_LAYER` freely (ideally as long as they don't interfere with the bitmask values of the `QRMatrix::M_*` constants). So you just need to re-implement the `QROutputAbstract::collectModules()` method to do this and then you can use these layers as CSS classes in the SVG defs. The example code for that would look like follows:

```php
<?php

use chillerlan\QRCode\Common\EccLevel;
use chillerlan\QRCode\Data\QRMatrix;
use chillerlan\QRCode\Output\QRMarkupSVG;
use chillerlan\QRCode\{QRCode, QROptions};

require_once __DIR__.'/../vendor/autoload.php';

class RandomDotsSVGOutput extends QRMarkupSVG{

	/**
	 * @inheritDoc
	 */
	protected function collectModules(Closure $transform):array{
		$paths = [];

		// collect the modules for each type
		foreach($this->matrix->matrix() as $y => $row){
			foreach($row as $x => $M_TYPE){
				$M_TYPE_LAYER = $M_TYPE;

				if($this->options->connectPaths 
				   && $this->matrix->checkTypeNotIn($x, $y, $this->options->excludeFromConnect)
				){
					// to connect paths we'll redeclare the $M_TYPE_LAYER to data only
					$M_TYPE_LAYER = QRMatrix::M_DATA;

					if($this->matrix->check($x, $y)){
						$M_TYPE_LAYER |= QRMatrix::IS_DARK;
					}
				}

				// randomly assign another $M_TYPE_LAYER for the given types
				// note that the layer id has to be an integer value,
				// ideally outside of the several bitmask values
				if($M_TYPE_LAYER === (QRMatrix::M_DATA | QRMatrix::IS_DARK)){
					$M_TYPE_LAYER = [1234, 5678][rand(0, 1)];
				}

				// collect the modules per $M_TYPE
				$paths[$M_TYPE_LAYER][] = $transform($x, $y, $M_TYPE, $M_TYPE_LAYER);
			}
		}

		// beautify output
		ksort($paths);

		return $paths;
	}

}


$options = new QROptions([
	'version'             => 5,
	'eccLevel'            => EccLevel::H,
	'addQuietzone'        => true,
	'imageBase64'         => false,
	'outputType'          => QRCode::OUTPUT_CUSTOM,
	'outputInterface'     => RandomDotsSVGOutput::class,
	'markupDark'          => '',
	'markupLight'         => '',
	'imageTransparent'    => false,

	'connectPaths'        => true,
	'excludeFromConnect'  => [
		QRMatrix::M_FINDER|QRMatrix::IS_DARK,
		QRMatrix::M_FINDER_DOT|QRMatrix::IS_DARK,
		QRMatrix::M_ALIGNMENT|QRMatrix::IS_DARK,
	],

	'drawCircularModules' => true,
	'circleRadius'        => 0.4,
	'keepAsSquare'        => [
		QRMatrix::M_FINDER|QRMatrix::IS_DARK,
		QRMatrix::M_FINDER_DOT|QRMatrix::IS_DARK,
		QRMatrix::M_ALIGNMENT|QRMatrix::IS_DARK,
	],

	// assign the newly created layers here
	'svgDefs'             => '
	<style><![CDATA[
		.light{ fill: #ddd; }
		.dark{ fill: #424242; }
		.qr-1234 { fill: #463ed7; }
		.qr-5678 { fill: #4a4699; }
	]]></style>',

]);


header('content-type: image/svg+xml');
echo (new QRCode($options))->render('https://www.youtube.com/watch?v=dQw4w9WgXcQ');
```

This is what the output looks like:
![random](https://user-images.githubusercontent.com/592497/178030014-d6bf5f07-a060-41c1-bf76-e37e012ec2bf.svg)

**Reply by codemasher** [reply]
I've posted the full (updated) example [over here](https://github.com/chillerlan/php-qrcode/blob/f6f98583599e34e7bd61f7656d301b79c68d9a94/examples/svgRandomColoredDots.php).

**Reply by sogorman** [reply]
This is fantastic! Will give it a look today!a
